### ISSUE TITLE
Questions about sigma search and epsilon from composed tradeoff functions

### ISSUE THREAD

**Issue by xplip** [lxuechen/private-transformers#8]
(Making a new issue for this because you probably weren't notified of my comment in the closed original issue)

Sorry for having to reopen this, but I do have two more (perhaps related) questions after all and would really appreciate if you could help clarify them.

1) When using the automated sigma search (based on a specified target epsilon and N epochs), the final epsilon computed by the `PrivacyEngine` after training for N epochs is always much higher than the target epsilon, so it seems that the sigma chosen by `get_sigma_from_rdp` is too high. This also happens when I run the sentence classification and the table2test examples in the repo. E.g., instead of my target epsilon 8, I will end up with something like epsilon 10-11. How did you get your final epsilon to match the target epsilon in the experiments in your paper?

2) How do you compute the converted epsilon from composed tradeoff functions when let's say training SST-2 with the default hyperparameters from the examples? Do you reduce the `num_compositions=1000` in `_eps_from_glw` to something way lower than 1000 because the script only runs for ~400 optimization steps and would otherwise always throw the `Numerical composition of tradeoff functions failed! Double check privacy parameters.` error?

_Originally posted by @xplip in https://github.com/lxuechen/private-transformers/issues/7#issuecomment-987020758_

**Comment by lxuechen** [comment]
Hi Phillip,

Thanks for the issue. Sorry about my slow response as it's the last week of school and I'm attending neurips. 

For 1, would you mind sharing some code snippets? I suspect what you're seeing is caused by the auto-search algorithm using `rdp_cks ` mode (this is the default). This mode still accounts privacy loss with RDP, but the RGP to approximate-DP conversion is tighter compared to the `rdp` mode. So the sigma obtained in the `rdp_cks` mode will be smaller than that obtained in the `rdp` mode. The `rdp_cks` mode is also what's being used as the default in Opacus right now. 

I think 2 is a bug, and I've fixed it in `main` just now. Thanks for spotting it! 

Hope this helps! Feel free to drop in other questions and/or issues!

**Comment by xplip** [comment]
Hi Chen, thanks for already fixing the glw accounting issue despite being busy! 

So, regarding 1) Based on what's happening in `get_privacy_spent()`, it seems that when I use `accounting_mode="rdp"` for the `PrivacyEngine`, I should look at `eps_rdp_opacus` in the output of `get_privacy_spent()` and when using  `accounting_mode="rdp_cks"`, I should look at `eps_rdp`. Now, in both cases I have encountered that the final epsilon was larger than my specified target epsilon. I've done some debugging today and I think I've narrowed down the problem (but I may just be using something wrong).

Each time I call `optimizer.step(loss=losses.get("vector_loss"))` in my training loop, the step counter in the `PrivacyEngine` is actually incremented twice, once in `step` and once in `dp_step` directly:

```
        # Override step.
        def dp_step(_self, **kwargs):
            closure = kwargs.pop("closure", None)

            _self.privacy_engine.step(**kwargs)
            _self.original_step(closure=closure)
            _self.privacy_engine.unlock()  # Only enable creating new grads once parameters are updated.
            _self.privacy_engine.steps += 1
```
```
  @torch.no_grad()
  def step(self, scale=1, **kwargs):
      """Step function.
      `loss` must be passed in as a keyword argument!
      """
      if self.ghost_clipping:
          self._ghost_step(loss=kwargs.pop("loss"))
      else:
          self._step(loss=kwargs.pop("loss"), scale=scale)
      self.steps += 1  # Make privacy accounting correct.
```

So then, whenever I call `get_privacy_spent` and rely on the `PrivacyEngine`'s step counter rather than passing the current training step as argument explicitly, the privacy spending is calculated for twice the actual step count (e.g., 1400 when my total number of training steps is actually 700). On the contrary, the sigma obtained through `get_sigma_from_rdp`/`get_sigma_from_rdp_cks` would be based on a step count of 700. Therefore, the more steps I train for, the further the calculated epsilon diverges from my target epsilon. It seems to be fixed when I comment out one of the `steps += 1`, but I'm not sure if this is intended behavior and I should make a change elsewhere in my code. Does this make sense?

**Comment by lxuechen** [comment]
Thanks for the catch, Phillip! 

This issue isn't present in my private repo, and must have gone in while refactoring. It's fixed now.
